packages/core-data/src/entity-provider.js

```diff
@@ -184,6 +190,45 @@ export function useEntityBlockEditor( kind, name, { id: _id } = {} ) {
 		}
 	}, [ content ] );
 
+	const updateFootnotes = useCallback(
+		( _blocks ) => {
+			const footnotes = meta?.footnotes
+				? JSON.parse( meta.footnotes )
+				: [];
+			const _content = getRichTextValues( _blocks ).join( '' ) || '';
+
+			if ( _content.indexOf( 'data-fn' ) === -1 ) return;
+
+			// This can be avoided when
+			// https://github.com/WordPress/gutenberg/pull/43204 lands. We can then
+			// get the order directly from the rich text values.
+			const regex = /data-fn="([^"]+)"/g;
+			const newOrder = [];
+			let match;
+			while ( ( match = regex.exec( _content ) ) !== null ) {
+				newOrder.push( match[ 1 ] );
+			}
+
+			const currentOrder = footnotes.map( ( fn ) => fn.id );
+
+			if ( currentOrder.join( '' ) === newOrder.join( '' ) ) return;
+
+			const newFootnotes = newOrder.map(
+				( fnId ) =>
+					footnotes.find( ( fn ) => fn.id === fnId ) || {
+						id: fnId,
+						content: '',
+					}
+			);
```

Comment:
There's probably some logic here that we should test properly.

Reply:
Added e2e tests

Reply:
Of course there's way more things we should be adding e2e tests for as we iterate on this